### PR TITLE
NEW Add method annotation checks

### PR DESCRIPTION
**Pull request by GuySartorelli**
Since this doesn't have CI yet, you'll have to check the CI in creative-commoners instead of on this PR directly.
See https://github.com/creative-commoners/silverstripe-standards/actions/workflows/ci.yml

Note that the PDO failure will be fixed by https://github.com/silverstripe/gha-ci/issues/97

## How to test this manually

1. Add this repo and `phpstan/extension-installer` as dev dependencies
2. Add a `phpstan.neon` or `phpstan.neon.dist` file to your project root with this content:
    ```neon
    parameters:
      paths:
        - vendor/silverstripe/framework/src/Security
    ```
3. Run `vendor/bin/phpstan`
    This will run it against the path set above, and due to the rules set in this repo it will only run our custom rule.
4. It should be pretty quick, and the only failures you should see should relate to `@method` annotations.

## Issue
- https://github.com/silverstripe/.github/issues/171

### CODE REVIEW

File: src/PHPStan/MethodAnnotationsRule.php [hunk 1]
```diff
@@ -0,0 +1,359 @@
...
+        foreach ($existingAnnotations as $methodName => $annotationData) {
+            $annotationString = $annotationData['annotationString'];
+            $returnString = $annotationData['returnString'];
+
+            // Complain about any @method annotations we don't expect
+            if (!array_key_exists($methodName, $expectedAnnotations)) {
+                $errors[] = RuleErrorBuilder::message(
+                    "@method annotation '$annotationString' isn't expected and should be removed."
+                )->build();
+                continue;
```
Comment (emteknetnz) [inline src/PHPStan/MethodAnnotationsRule.php:71]:
```suggestion
            // Complain about any @method annotations we don't expect
            $failed = false;
            if (!array_key_exists($methodName, $expectedAnnotations)) {
                $errors[] = RuleErrorBuilder::message(
                    "@method annotation '$annotationString' isn't expected and should be removed."
                )->build();
                $failed = true;
```
And then at the end of all these checks go

```php
if ($failed) {
    continue;
}
```

Means that multiple errors get captured in one go

File: src/PHPStan/MethodAnnotationsRule.php [hunk 2]
```diff
@@ -0,0 +1,359 @@
...
+            }
+
+            // Complain about @method annotations if a real method exists with that name
+            if ($classReflection->hasNativeMethod($methodName)) {
```
Comment (emteknetnz) [inline src/PHPStan/MethodAnnotationsRule.php:75]: Does `hasNativeMethod()` include inherited methods?

**Review by emteknetnz — CHANGES_REQUESTED** [review]
Should also validate the that classes like HasManyList are imported when they're used for method annotations

**Comment by GuySartorelli** [comment]
> Should also validate the that classes like HasManyList are imported when they're used for method annotations

Do you mean checking for missing `use` statements? That's done already, see the code under the "Complain about missing use statements" comment.

File: src/PHPStan/MethodAnnotationsRule.php (hunk 1 above)
Comment (ChloeSartorelli) [reply src/PHPStan/MethodAnnotationsRule.php:71]: The continue statements are there intentionally to only include relevant errors.

- If we don't expect the annotation, the only error related to it should be to remove it.
- If a method exists for an annotation, the only error related to it should be to remove it.
- If an annotation is duplicated, we won't know what other errors should be included until the duplicate is removed.
- If there's a missing use statement, errors about type mismatches will be wrong - the type isn't mismatched, it's just missing a use statement.
- If there's a type mismatch, that's why it's wrong. We don't want to report the same error twice.

File: src/PHPStan/MethodAnnotationsRule.php (hunk 2 above)
Comment (ChloeSartorelli) [reply src/PHPStan/MethodAnnotationsRule.php:75]: Yup

**Review by emteknetnz — CHANGES_REQUESTED** [review]
Tested locally with steps in description, though I tried running on all of framework rather than just the Security dir.  I got this:

```
Internal error: Internal error: Too few arguments to function                                  
SilverStripe\ORM\Hierarchy\Hierarchy::get_extra_config(), 0 passed in                          
/path/to/somesite/vendor/silverstripe/standards/src/PHPStan/MethodAnnotationsRule.php  
on line 320 and exactly 3 expected while analysing file                                        
/path/to/somesite/vendor/silverstripe/framework/src/ORM/Hierarchy/Hierarchy.php        
Run PHPStan with -v option and post the stack trace to:                                        
https://github.com/phpstan/phpstan/issues/new?template=Bug_report.yaml                         
Child process error (exit code 1):                                                             
Internal error: Internal error: Internal error. while analysing file                           
/path/to/somesite/vendor/silverstripe/framework/src/ORM/PaginatedList.php              
Run PHPStan with -v option and post the stack trace to:                                        
https://github.com/phpstan/phpstan/issues/new?template=Bug_report.yaml                         
Child process error (exit code 1):                                                             
Internal error: Internal error: Internal error. while analysing file                           
/path/to/somesite/vendor/silverstripe/framework/src/ORM/ListDecorator.php              
Run PHPStan with -v option and post the stack trace to:                                        
https://github.com/phpstan/phpstan/issues/new?template=Bug_report.yaml                         
Child process error (exit code 1):                                                             
Internal error: Internal error: Internal error. while analysing file                           
/path/to/somesite/vendor/silverstripe/framework/src/ORM/Connect/MySQLDatabase.php      
Run PHPStan with -v option and post the stack trace to:                                        
https://github.com/phpstan/phpstan/issues/new?template=Bug_report.yaml                         
Internal error: Internal error: Too few arguments to function                                  
SilverStripe\ORM\Search\FulltextSearchable::get_extra_config(), 0 passed                       
in                                                                                             
/path/to/somesite/vendor/silverstripe/standards/src/PHPStan/MethodAnnotationsRule.php  
on line 320 and exactly 3 expected while analysing file                                        
/path/to/somesite/vendor/silverstripe/framework/src/ORM/Search/FulltextSearchable.php  
Run PHPStan with -v option and post the stack trace to:                                        
https://github.com/phpstan/phpstan/issues/new?template=Bug_report.yaml                         
Internal error: Internal error: Internal error. while analysing file                           
/path/to/somesite/vendor/silverstripe/framework/src/ORM/GroupedList.php                
Run PHPStan with -v option and post the stack trace to:                                        
https://github.com/phpstan/phpstan/issues/new?template=Bug_report.yaml                         
Child process error (exit code 1):      
```

**Comment by GuySartorelli** [comment]
Ahh, yes, looks like in my haste I didn't notice that the method signature for `get_extra_config()` actually takes arguments which can change the returned value. This will be interesting.

**Comment by GuySartorelli** [comment]
Turns out the method is ignored if it's declared on a `DataObject` directly.
When it's declared on an `Extension` subclass, it has this signature:
```php
/**
 * @param string $class The FQCN of the class being extended (will always be DataObject from PHPStan's pov)
 * @param string $extension The FQCN of this class
 * @param array $args Extra arguments as part of the extension configuration (will always be an empty array from PHPStan's pov)
 */
public static function get_extra_config($class, $extension, $args): array;
```

That's accounted for now.

**Comment by GuySartorelli** [comment]
The remaining errors don't seem to be related to the custom rule. They occur before any of the code in this repository is executed. Those errors are:

```text
Internal error: Internal error: Internal error. while analysing file                                
     /home/gsartorelli/projects/phpstan/vendor/silverstripe/framework/src/ORM/Connect/MySQLDatabase.php  
     Run PHPStan with -v option and post the stack trace to:                                             
     https://github.com/phpstan/phpstan/issues/new?template=Bug_report.yaml                              
     Child process error (exit code 1):                                                                  
     Internal error: Internal error: Internal error. while analysing file                                
     /home/gsartorelli/projects/phpstan/vendor/silverstripe/framework/src/ORM/GroupedList.php            
     Run PHPStan with -v option and post the stack trace to:                                             
     https://github.com/phpstan/phpstan/issues/new?template=Bug_report.yaml                              
     Internal error: Internal error: Internal error. while analysing file                                
     /home/gsartorelli/projects/phpstan/vendor/silverstripe/framework/src/ORM/PaginatedList.php          
     Run PHPStan with -v option and post the stack trace to:                                             
     https://github.com/phpstan/phpstan/issues/new?template=Bug_report.yaml                              
     Internal error: Internal error: Internal error. while analysing file                                
     /home/gsartorelli/projects/phpstan/vendor/silverstripe/framework/src/ORM/ListDecorator.php          
     Run PHPStan with -v option and post the stack trace to:                                             
     https://github.com/phpstan/phpstan/issues/new?template=Bug_report.yaml                              
     Child process error (exit code 1):
```
I've raised https://github.com/phpstan/phpstan/issues/10509 about that, but IMO that shouldn't be a blocker for merging this PR (though it will be a blocker for adding the PHPstan config to the framework repo).

**Review by emteknetnz — APPROVED** [review]

**Pull request merged**